### PR TITLE
PHPUnit on GitHub Actions

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

Continuing #41664

### CODE REVIEW

File: .github/workflows/unit-tests.yml [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+            composer require --dev --no-update masterminds/html5:~2.7.5@dev
+            composer update --no-progress --ansi
+          echo "::endgroup::"
+
+          echo "::group::install phpunit"
+            ./phpunit install
+          echo "::endgroup::"
+
+      - name: Patch return types
+        if: "${{ matrix.php == '8.0' }}"
```
Comment (stof) [inline .github/workflows/unit-tests.yml]: shouldn't we also patch them on PHP 8.1 ?

File: .travis.yml [hunk 2]
```diff
@@ -109,22 +107,13 @@ before_install:
...
-          (cd php-$MIN_PHP && ./configure --enable-sigchild --enable-pcntl && make -j2)
-      fi
-
     - |
       # php.ini configuration
       (
-          for PHP in $TRAVIS_PHP_VERSION $php_extra; do
-              ([[ $PHP != 7.4 ]] && phpenv global $PHP 2>/dev/null) || (cd / && wget https://storage.googleapis.com/travis-ci-language-archives/php/binaries/ubuntu/18.04/x86_64/php-$PHP.tar.bz2 -O - | tar -xj) &
-          done
+          ([[ $TRAVIS_PHP_VERSION != 7.4 ]] && phpenv global $TRAVIS_PHP_VERSION 2>/dev/null) || (cd / && wget https://storage.googleapis.com/travis-ci-language-archives/php/binaries/ubuntu/18.04/x86_64/php-$TRAVIS_PHP_VERSION.tar.bz2 -O - | tar -xj) &
```
Comment (stof) [inline .travis.yml]: do we still need this custom download ? We don't use multiple PHP versions anymore.

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -1,4 +1,4 @@
-name: Tests
+name: PHPUnit
```
Comment (bendavies) [inline .github/workflows/tests.yml:1]: probably want 
```yaml
defaults:
  run:
    shell: bash
```
in this file so your bash scripts fail appropriately

File: .github/workflows/unit-tests.yml [hunk 4]
```diff
@@ -0,0 +1,210 @@
+name: PHPUnit
```
Comment (bendavies) [inline .github/workflows/unit-tests.yml:1]: probably want 
```yaml
defaults:
  run:
    shell: bash
```
in this file so your bash scripts fail appropriately

File: .github/workflows/unit-tests.yml (hunk 1 above)
Comment (nicolas-grekas) [reply .github/workflows/unit-tests.yml]: we just need one job to check return types

**Review by nicolas-grekas — COMMENTED** [review]
I think I'm done with this PR, except for the failures, which need some investigation.

TL;DR, everything is moved to GHA, except tty and zookeeper tests:
- GHA is not compatible with tty mode;
- zookeeper is not included in our integration tests.

I would appreciate some help to fix the failures if anyone wants to have a look!

**Comment by wouterj** [comment]
Fyi, I just pushed 6713ab3291b3ce41eb7ab5d702000239a460c1e7 to 4.4 to fix the security XML failure (the branch was broken)

**Comment by nicolas-grekas** [comment]
Greeeen! :green_apple:

**Comment by wouterj** [comment]
Greeeat! Thanks a lot of effort @nicolas-grekas and @derrabus! I also love the speed improvement over Travis :)

Fyi: Looking at all the stuff we do, I think it would be good to make some sort of overview of all types of test we run (both to increase the bus factor of our CI configuration and to help contributors discover why things fail). I'll try to come up with a list of test cases I'm aware of in a doc PR shortly :)

**Comment by nicolas-grekas** [comment]
Please do!
Quickly (for 4.4):
- 7.1 + x86 + with&without extensions + Windows on appveyor 
- 7.2, 8.0 on GHA
- 7.4+high-deps, 8.0+low-deps on GHA
- 8.0 flipped tests on GHA (aka checkout previous version and test with patched deps)
- 7.3 + tty + zookeeper on Travis-ci

**Comment by nicolas-grekas** [comment]
> fatal: couldn't find remote ref 3.4

new transient behavior :)

**Comment by nicolas-grekas** [comment]
Thank you @derrabus.

**Comment by derrabus** [comment]
Thank you for completing the PR, @nicolas-grekas! Its good to have a proper CI again. 🚀

File: .github/workflows/unit-tests.yml [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+              git checkout -m FETCH_HEAD
+              COMPONENTS=$(echo "$COMPONENTS" | xargs dirname | xargs -n1 -I{} bash -c "[ -e '{}/phpunit.xml.dist' ] && echo '{}'" | sort || true)
+              (cd src/Symfony/Component/HttpFoundation; composer require --dev --no-update mongodb/mongodb)
+              [[ ! $COMPONENTS ]] || echo "$COMPONENTS" | parallel -j +3 "_run_tests {} 'cd {} && rm composer.lock vendor/ -Rf && $COMPOSER_UP && $PHPUNIT$LEGACY'" || X=1
+          fi
+
+          [[ ! $X ]] || (exit 1)
+
+      - name: Run tests with SIGCHLD enabled PHP
+        if: "${{ matrix.php == '7.1' }}"
```
Comment (derrabus) [inline .github/workflows/unit-tests.yml:206]: @nicolas-grekas You have removed PHP 7.1 from the matrix. Because of that, this step is not executed at the moment.

Comment (nicolas-grekas) [reply .github/workflows/unit-tests.yml:206]: Right! Fixed in bb59651e48

Comment (derrabus) [reply .github/workflows/unit-tests.yml:206]: Thanks!
